clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3631,6 +3698,38 @@ SYCLIntelMaxWorkGroupSizeAttr *Sema::MergeSYCLIntelMaxWorkGroupSizeAttr(
       return nullptr;
   }
 
+  // If the [[intel::max_work_group_size(X, Y, Z)]] attribute is specified on
+  // a declaration along with [[sycl::reqd_work_group_size(X1, Y1, Z1)]]
+  // attribute, check to see if values of reqd_work_group_size arguments are
+  // equal or less than values of max_work_group_size attribute arguments.
+  //
+  // The arguments to reqd_work_group_size are ordered based on which index
+  // increments the fastest. In OpenCL, the first argument is the index that
+  // increments the fastest, and in SYCL, the last argument is the index that
+  // increments the fastest.
+  //
+  // [[sycl::reqd_work_group_size]] and [[cl::reqd_work_group_size]] are
+  // available in SYCL modes and follow the SYCL rules.
+  // __attribute__((reqd_work_group_size)) is only available in OpenCL mode
+  // and follows the OpenCL rules.
+  if (const auto *DeclAttr = D->getAttr<ReqdWorkGroupSizeAttr>()) {
+    bool checkFirstArgument = getLangOpts().OpenCL
+	      ? checkWorkGroupSizeAttrValues(DeclAttr->getXDim(), A.getZDim())
+	      : checkWorkGroupSizeAttrValues(DeclAttr->getXDim(), A.getXDim());
+    bool checkSecondArgument =
+	      checkWorkGroupSizeAttrValues(DeclAttr->getYDim(), A.getYDim());
+    bool checkThirdArgument = getLangOpts().OpenCL
+	      ? checkWorkGroupSizeAttrValues(DeclAttr->getZDim(), A.getXDim())
+	      : checkWorkGroupSizeAttrValues(DeclAttr->getZDim(), A.getZDim());
```

Comment:
```suggestion
    bool CheckFirstArgument = getLangOpts().OpenCL
	      ? checkWorkGroupSizeAttrValues(DeclAttr->getXDim(), A.getZDim())
	      : checkWorkGroupSizeAttrValues(DeclAttr->getXDim(), A.getXDim());
    bool CheckSecondArgument =
	      checkWorkGroupSizeAttrValues(DeclAttr->getYDim(), A.getYDim());
    bool CheckThirdArgument = getLangOpts().OpenCL
	      ? checkWorkGroupSizeAttrValues(DeclAttr->getZDim(), A.getXDim())
	      : checkWorkGroupSizeAttrValues(DeclAttr->getZDim(), A.getZDim());
```
Now that I've seen this logic three times, I sort of wonder whether we should add a helper function to test the entire predicate.

Reply:
>>Now that I've seen this logic three times, I sort of wonder whether we should add a helper function to test the entire predicate.

Do you have any suggestion for this, @AaronBallman? i thought about it. checkWorkGroupSizeAttrValues() is a helper function here. We are accessing max_work_group_size() attribute values differently three times, that's the concern i have.

Reply:
Hmmm, yeah, one of these ways is different than the other two, so there's not as much to be gained as I thought. We could combine two of them, but I'm not certain it's worth the effort.

Reply:
>>We could combine two of them, but I'm not certain it's worth the effort.
Both are different, so combining is not possible.
checkWorkGroupSizeAttrValues(DeclAttr->getXDim(), A.getZDim())
checkWorkGroupSizeAttrValues(DeclAttr->getXDim(), ZDim)

Reply:
I was thinking that both of those are comparing an `Expr *` to another `Expr *`. Or does one of those return an `APValue` and the other returns an `Expr *`?

Reply:
> I was thinking that both of those are comparing an `Expr *` to another `Expr *`. Or does one of those return an `APValue` and the other returns an `Expr *`?

>>does one of those return an `APValue` and the other returns an `Expr *`?
Yes, correct. reqd_work_group_size attribute is not refactored yet.

Reply:
Aha, then yeah, definitely no reason to try to combine. :-) Thanks!